### PR TITLE
Add a HEAD route for every GET

### LINKED ISSUE #1489: Automatically provide HEAD for GET
### Expected Behaviour

Similarly to how it works in Spring, it would be useful if every Get-method in a controller automatically had a Head counterpart, since Head is essentially Get without the body.

E.g. once this is defined:

```
@Controller("/v1")
@Validated
public class RootResource {

  @Produces("application/json")
  @Get(
      uri = "/{id}"
  )
  public MyModel getEntity(String id) {
    return ...;
  }
}
```

I can perform both operations:
`GET /v1/1234`
`HEAD /v1/1234`

### Actual Behaviour

Only `GET /v1/1234` works, a HEAD controller method would need to be specified manually.

### PR DESCRIPTION
**Pull request by jameskleeh**
Closes #1489.

If you know of a better way of going about this let me know. I tried to do it in the abstract route builder itself but since only a single route is returned, the caller would have no way of further processing the route.

### CODE REVIEW

File: function-web/src/main/java/io/micronaut/function/web/AnnotatedFunctionRouteBuilder.java [hunk 1]
```diff
@@ -94,7 +94,7 @@ public void process(BeanDefinition<?> beanDefinition, ExecutableMethod<?, ?> met
             String functionName = beanDefinition.getValue(FunctionBean.class, String.class).orElse(methodName);
             String functionMethod = beanDefinition.getValue(FunctionBean.class, "method", String.class).orElse(null);
 
-            UriRoute route = null;
+            List<UriRoute> routes = new ArrayList<>();
```
Comment (graemerocher) [inline function-web/src/main/java/io/micronaut/function/web/AnnotatedFunctionRouteBuilder.java]: Specify size of `2` to the array

Comment (jameskleeh) [reply function-web/src/main/java/io/micronaut/function/web/AnnotatedFunctionRouteBuilder.java]: Done

**Comment by graemerocher** [comment]
Failing tests?

**Comment by jameskleeh** [comment]
Looks to be set ordering funny business:

```
    json.message == 'Method [POST] not allowed. Allowed methods: [GET, HEAD]'
    |    |       |
    |    |       false
    |    |       6 differences (89% similarity)
    |    |       Method [POST] not allowed. Allowed methods: [(H)E(AD), (G)E(T-)]
    |    |       Method [POST] not allowed. Allowed methods: [(G)E(T-), (H)E(AD)]
    |    Method [POST] not allowed. Allowed methods: [HEAD, GET]
```

**Pull request merged**